pwiz_tools/Skyline/SettingsUI/PeptideSettingsUI.cs

```diff
@@ -728,65 +732,259 @@ public void ShowBuildLibraryDlg()
+
+                            builderLite.PreferEmbeddedSpectra = response == UpdateProgressResponse.normal;
+
+                            retry = true;
+                            continue;
+                        }
+                        else
+                        {
+                            MessageDlg.ShowException(this, status.ErrorException);
                         }
-                        return;
                     }
+                    retry = false;
+                }
+            } while (retry);
 
-                    IsBuildingLibrary = true;
+            CompleteLibraryBuild(buildState);
 
-                    var builder = dlg.Builder;
+            Settings.Default.SpectralLibraryList.Add(builder.LibrarySpec);
+            _driverLibrary.LoadList();
+            var libraryIndex = listLibraries.Items.IndexOf(builder.LibrarySpec.Name);
+            if (libraryIndex >= 0)
+                listLibraries.SetItemChecked(libraryIndex, true);
 
-                    // assume success and cleanup later
-                    Settings.Default.SpectralLibraryList.Add(builder.LibrarySpec);
-                    _driverLibrary.LoadList();
-                    var libraryIndex = listLibraries.Items.IndexOf(builder.LibrarySpec.Name);
-                    if (libraryIndex >= 0)
-                        listLibraries.SetItemChecked(libraryIndex, true);
+            IsBuildingLibrary = false;
+        }
 
-                    var currentForm = this;
+        public void CompleteLibraryBuild(BuildState buildState)
+        {
+            if (!string.IsNullOrEmpty(buildState.ExtraMessage))
+            {
+                MessageDlg.Show(this, buildState.ExtraMessage);
+            }
 
-                    _libraryManager.BuildLibrary(_parent, builder, (buildState, success) =>
+            if (buildState.IrtStandard != null && !buildState.IrtStandard.IsEmpty)
+            {
+                // Load library
+                Library lib = null;
+                using (var longWait = new LongWaitDlg())
+                {
+                    longWait.Text = SettingsUIResources.LibraryBuildNotificationHandler_AddIrts_Loading_library;
+                    var status = longWait.PerformWork(this, 800, monitor =>
                     {
-                        _parent.LibraryBuildCompleteCallback(buildState, success);
+                        lib = _libraryManager.TryGetLibrary(buildState.LibrarySpec) ??
+                              _libraryManager.LoadLibrary(buildState.LibrarySpec, () => new DefaultFileLoadMonitor(monitor));
+                        if (lib != null)
+                        {
+                            foreach (var stream in lib.ReadStreams)
+                                stream.CloseStream();
+                        }
+                    });
+                    if (status.IsCanceled)
+                        lib = null;
+                    if (status.IsError)
+                        throw status.ErrorException;

```

Comment:
I don't think there is anyone on the callstack who is going to catch this exception. This is going to end up being an unhandled exception.

This should probably be:
`MessageDlg.ShowException(status.ErrorException);`

If you were going to throw the exception, then I think it should be:
`Helpers.WrapAndThrowException(status.ErrorException);`
so that the callstack is preserved.

Reply:
Funny. This pattern with LongWaitDlg was repeated 4 more times in the function below and then some extra exception handling for the last 2 times that added a more specific message. I think I cleaned everything up with 2 new local utility functions. Please have a look at that. It does make me wonder how all of this was handled inside LibraryBuildNotification, and it gives me the strong feeling that the error handling here is never tested, since nothing failed when there would have been unhandled exceptions in the error cases.